File: leetcode_problems/typescript/src/lc1560_most_visited_sector_in_circular_track.ts
function mostVisited(n: number, rounds: number[]): number[] {
  let arr: number[] = new Array(n).fill(0);

  for (let i = 1; i < rounds.length; i++) {
    let start = rounds[i - 1];
    let end = rounds[i];
    if (start < end) {
      for (let j = start; j < end; j++) {
        arr[j - 1]++;
      }
    } else {
      for (let j = start; j <= n; j++) {
        arr[j - 1]++;
      }
      for (let j = 0; j < end; j++) {
        arr[j - 1]++;
      }
    }
  }
  arr[rounds[rounds.length - 1] - 1]++;

  let xmax: number = 0;
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] > xmax) {
      xmax = arr[i];
    }
  }

  let res: number[] = [];
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] == xmax) {
      res.push(i + 1);
    }
  }

  return res;
}

function mostVisited2(n: number, rounds: number[]): number[] {
  const start = rounds[0];
  const end = rounds[rounds.length - 1];
  const result: number[] = [];

  if (start <= end) {
    for (let i = start; i <= end; i++) {
      result.push(i);
    }
  } else {
    for (let i = 1; i <= end; i++) {
      result.push(i);
    }
    for (let i = start; i <= n; i++) {
      result.push(i);
    }
  }

  return result;
}

console.log(mostVisited(4, [1, 3, 1, 2]));

File: leetcode_problems/cpp/lc1560_most_visited_sector_in_circular_track.cpp
#include <iostream>
#include <vector>
using std::vector;

class Solution
{
public:
    vector<int> mostVisited(int n, vector<int> &rounds)
    {
        vector<int> arr(n, 0);
        for (int i = 1; i < rounds.size(); i++)
        {
            int start = rounds[i - 1];
            int end = rounds[i];
            if (start <= end)
            {
                for (int j = start; j < end; j++)
                {
                    arr[j - 1]++;
                }
            }
            else
            {
                for (int j = start; j <= n; j++)
                {
                    arr[j - 1]++;
                }
                for (int j = 1; j < end; j++)
                {
                    arr[j - 1]++;
                }
            }
        }
        arr[rounds.back() - 1]++;

        int xmax = 0;
        for (int i = 0; i < n; i++)
        {
            if (arr[i] > xmax)
            {
                xmax = arr[i];
            }
        }

        vector<int> res;
        for (int i = 0; i < n; i++)
        {
            if (arr[i] == xmax)
            {
                res.push_back(i + 1);
            }
        }
        return res;
    }

    vector<int> mostVisited2(int n, vector<int> &rounds)
    {
        const int start = rounds[0];
        const int end = rounds[rounds.size() - 1];
        vector<int> res;
        if (start <= end)
        {
            for (int i = start; i <= end; i++)
            {
                res.push_back(i);
            }
        }
        else
        {
            for (int i = 1; i <= end; i++)
            {
                res.push_back(i);
            }
            for (int i = start; i <= n; i++)
            {
                res.push_back(i);
            }
        }
        return res;
    }
};

int main()
{
    vector<int> rounds = {1, 3, 1, 2};
    int n = 4;
    Solution sol;
    auto res = sol.mostVisited(n, rounds);
    for (int i = 0; i < res.size(); i++)
    {
        std::cout << res[i] << std::endl;
    }
    return 0;
}
